### DISCUSSION TITLE
Why ClassGraph finds classes from other deployment on Wildfly?

### DISCUSSION THREAD

**Discussion by mkuligPSI** [classgraph/classgraph#700] [Q&A]
I prepared WAR archive and I deployed it on Wildfly 21. 
I want to find all classes annotated with` javax.ws.rs.ApplicationPath` annotation. 

```
try (ScanResult scanResult = new ClassGraph().enableAllInfo().whitelistPackages( PACKAGE ).scan())
        {
            final ClassInfoList classesWithAnnotation =
                    scanResult.getClassesWithAnnotation( ApplicationPath.class.getName() );
        }
```
It finds two classes, one from my deployment and one from other deployment. I undeployed other deployment to make sure that one class is taken from it and I am sure, if only one deployment is deployed on Wildfly then it return only one class. 

I do not understand why ClassGraph finds classes from other deployment. How to configure ClassGraph  to scan only deployment  which ClassGraph itself belongs? I guess that I have to configure class loader, but I do not know how to do it correctly.

**Comment by lukehutch** [comment]
ClassGraph goes out of its way to try to find every scannable class in every classloader. I don't know what a deployment is in WildFly, but if it doesn't make any sense that classes in other deployments should be scannable, then we should change this behavior. Please look at the following classes, find the one that is matching your exact classloader, and try modifying it to find only your deployment.

https://github.com/classgraph/classgraph/tree/latest/src/main/java/nonapi/io/github/classgraph/classloaderhandler

**Reply by mkuligPSI** [reply]
There is [JBossClassLoaderHandler#findClasspathOrder](https://github.com/classgraph/classgraph/blob/e81f793faba16ed102aeba4760fa0b5cd6e260d2/src/main/java/nonapi/io/github/classgraph/classloaderhandler/JBossClassLoaderHandler.java#L208) method.
First `org.jboss.modules.Module` is retrieved from class loader. Next `getCallerModuleLoader` method is invoked and returns different class loader. Original class loader is an instance of  `org.jboss.modules.ModuleClassLoader` class. Class loader retrieved by `getCallerModuleLoader`  method is an instance of `org.jboss.as.server.moduleservice.ServiceModuleLoader` class. Next `moduleMap` method is invoked and returns map of modules.  This map contains both my deployments. Next operations are executed for each map entry. 

I was hoping that ClassGraph can be configured to meet my requirements. Unfortunately I have to modify `JBossClassLoaderHandler`. I do not have time to wait for official release and I do not want o to maintain my own version of `classgraph`.  I also do not see possibility to add my own Class Loader Handler without code change because handlers are kept in unmodifiable final filed in `ClassLoaderHandlerRegistry` class. I am afraid that I have to abandon ClassGraph  library.

**Comment by lukehutch** [comment]
If this is a genuine bug, I try to push bug fixes out within 24 hours.

Please try modifying the class to fix the issue, and submit a pull request. Please make the minimal change you can to fix the issue, since a lot of people rely on ClassGraph in JBoss, and I don't want to break things for people. Are you confident that nobody else would ever want to scan classes in other deployments, either?

**Reply by mkuligPSI** [reply]
I am not experienced with ClassGraph and I do not know how people use it in jboss.  It seems strange to me that ClassGraph is widely used in jboss and ClassGraph seraches classes in all deployments.  To be honest I used org.refelctions:reflections and because of one critical bug I considered migration to ClassGraph .  Fortunately, I have dealt with reflections issue so I abandon migration to ClassGraph . I do not feel enough competent and familiar with ClassGraph to provide changes. 
Thank you for help.

**Comment by lukehutch** [comment]
You can compile and install a local copy of ClassGraph by doing `./mvnw install`

**Comment by lukehutch** [comment]
Well I don't even k ow what a deployment is in JBoss, or how to replicate your issue. Please submit a complete minimal replication test case, and explain what you observe and what you expect. Then I can take a look. Thanks.
